File: Add Two Numbers LC-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        new = ListNode()
        new_head=new
        carry=0
        
        while l1 or l2 or carry:
          new.next=ListNode()
          new=new.next
          
          add=0          
          if l1:
            add=add+l1.val
            l1=l1.next
          if l2:
            add=add+l2.val
            l2=l2.next
          if carry:
            add+=carry
            carry=0
          
          new.val=add%10
          carry=add//10
            

        return new_head.next

File: Add Two Numbers LC-2.cpp
// tc=O(max(n,m))
// sc=o(1)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry=0;
        ListNode* head=new ListNode();
        ListNode *temp=head;

        while(l1 || l2 || carry){
            temp->next=new ListNode();
            temp=temp->next;
            
            int add=carry;
            add+=l1?l1->val:0;
            add+=l2?l2->val:0;
            
            temp->val=add%10;
            carry=add/10;
            
            
            if(l1) l1=l1->next;
            if(l2) l2=l2->next;
        }
        return head->next;
    }
};